inkcpp/include/globals.h

```diff
@@ -4,15 +4,43 @@
 
 namespace ink::runtime
 {
+	template<typename T>
+	struct ret {
+		operator T() { return value; }
+		operator bool() { return has_value; }
+		T value;
+		bool has_value;
+	};
 	/**
 	* Represents a global store to be shared amongst ink runners.
 	* Stores global variable values, visit counts, turn counts, etc.
 	*/
-	class globals_interface
+	class globals
 	{
 	public:
 		// No public interface yet
 		virtual void dummy() = 0;
-		virtual ~globals_interface() = default;
+		virtual ~globals() = default;
+
+		template<typename T>
+		ret<T> get(const char* name) {
```

Comment:
I would actually make this interface something like

```cpp
template<typename T>
bool get(const char* name, T& result) { ... }
```

This way we don't have to deal with cast operators (which I find sometimes don't do what you expect) AND the user doesn't need to specify the template type as it'll be inferred by the reference argument.

Reply:
The problem is that the user can't edit the Variable then.
Maybe then:
```cpp
template<typename T>
bool getPtr(const char* name, T* result) { ... }

template<typename T>
bool get(const char* name, T* result) const { ... } // for const access
```

Reply:
Oh, right. I didn't even think about writing... Something along the lines of your original solution is probably best then. Maybe make that templated `ret` struct a class, turn `value` into a pointer, and override the assignment operator?

Reply:
We can also just return a pointer
```cpp
template<typename T>
T* get(const char* name) { ... }
```
Then we don't need an extra class/struct.

Reply:
Oh, yeah, good point. `nullptr` means no variable found so that all works!